src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs

```diff
@@ -90,9 +120,85 @@ public async Task ProcessEventsAsync(IEnumerable<IEventMessage> events, string t
                 }
             }
 
-            // Step 3: Transform the Measurement Groups into Observations and write to the FHIR Server.
-            //         Exceptions are currently handled in ProcessMeasurementGroups
-            await ProcessMeasurementGroups(measurementGroups, template, log, measurementToEventMapping).ConfigureAwait(false);
+            // Transform the Measurement Groups into Observations and write to the FHIR Server.
+            // Exceptions are currently handled in SendMeasurementGroups
+            await SendMeasurementGroups(measurementGroups, template, log, measurementToEventMapping).ConfigureAwait(false);
+        }
+
+        private async Task ProcessMeasurementGroupEvents(IEnumerable<IEventMessage> events, ILookupTemplate<IFhirTemplate> template, ITelemetryLogger log)
+        {
+            if (events.Count() == 0)
+            {
+                return;
+            }
+
+            foreach (var evt in events)
+            {
+                try
+                {
+                    IEnumerable<MeasurementGroup> groupedMeasurements;
+
+                    try
+                    {
+                        var partitionId = events.FirstOrDefault()?.PartitionId;
+
+                        // decompress the measurement group if it is compressed
+                        byte[] bytes;
+                        if (evt.BodyContentType == "application/gzip")
+                        {
+                            bytes = Compression.DecompressWithGzip(evt.Body.ToArray());
+                            evt.Body = new ReadOnlyMemory<byte>(bytes);
+                        }
+                        else
+                        {
+                            bytes = evt.Body.ToArray();
+                        }
+
+                        // deserialize
+                        IEnumerable<Measurement> measurementGroup = JsonConvert.DeserializeObject<IEnumerable<Measurement>>(System.Text.Encoding.UTF8.GetString(bytes));

```

Comment:
Is there an option with compression that we can get a stream instead of a byte array?  We can get the readonly memory stream from event body.  So if we can use a stream and instead of byte we should be able to avoid the array allocation for the byte array.

Reply:
yeah we can use a stream. updated